presto-main/src/main/java/com/facebook/presto/operator/MergeSortProcessor.java

```diff
@@ -0,0 +1,191 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator;
+
+import com.facebook.presto.spi.Page;
+import com.google.common.collect.ImmutableList;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+
+import javax.annotation.Nullable;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.util.concurrent.Futures.immediateFuture;
+import static java.util.Objects.requireNonNull;
+
+public class MergeSortProcessor
+{
+    private static final ListenableFuture<?> NOT_BLOCKED = immediateFuture(null);
+
+    private final MergeSortComparator comparator;
+    private final List<MergeSource> mergeSources;
+
+    private ListenableFuture<?> blocked;
+
+    public MergeSortProcessor(MergeSortComparator comparator, List<? extends PageSupplier> pageSuppliers)
+    {
+        this.comparator = requireNonNull(comparator, "comparator is null");
+        requireNonNull(pageSuppliers, "pageSuppliers is null");
+        ImmutableList.Builder<MergeSource> mergeSources = ImmutableList.builder();
+        for (PageSupplier pageSupplier : pageSuppliers) {
+            mergeSources.add(new MergeSource(pageSupplier));
+        }
+        this.mergeSources = mergeSources.build();
+    }
+
+    @Nullable
+    public PageWithPosition poll()
+    {
+        PageWithPosition result = null;
+        for (MergeSource mergeSource : mergeSources) {
+            PageWithPosition current = mergeSource.getPage();
+            if (current == null) {
+                if (mergeSource.isFinished()) {
+                    continue;
+                }
+                else {
+                    return null;
+                }
+            }
+            if (result == null) {
+                result = current;
+            }
+            else {
+                int compareResult = comparator.compareTo(current.getPage(), current.getPosition(), result.getPage(), result.getPosition());
+                if (compareResult < 0) {
+                    result = current;
+                }
+            }
+        }
+        return result;
+    }
+
+    public ListenableFuture<?> isBlocked()
+    {
+        if (blocked != null && !blocked.isDone()) {
+            return blocked;
+        }
+
+        if (mergeSources.stream().allMatch(source -> source.isBlocked().isDone())) {
+            return NOT_BLOCKED;
+        }
+
+        blocked = Futures.allAsList(mergeSources.stream().map(MergeSource::isBlocked).collect(Collectors.toList()));
+        return blocked;
+    }
+
+    public boolean isFinished()
+    {
+        return mergeSources.stream().allMatch(MergeSource::isFinished);
+    }
+
+    private static class MergeSource
```

Comment:
It doesn't account for memory, see: `com.facebook.presto.operator.MergeHashSort.SingleChannelPagePositions`